api/types/replacement.go

```diff
@@ -22,6 +22,6 @@ type ReplSource struct {
 
 // ReplTarget defines where a substitution is to.
 type ReplTarget struct {
-	ObjRef    *Selector `json:"objref,omitempty" yaml:"objref,omitempty"`
-	FieldRefs []string  `json:"fieldrefs,omitempty" yaml:"fieldrefs,omitempty"`
+	ObjRef    *Target  `json:"objref,omitempty" yaml:"objref,omitempty"`
```

Comment:
This change implies slightly different fields, right? `Selector` has fields for label and annotation selection support, and `Target` has `APIVersion` instead. Either way, it seems to me this transformer is going to be more powerful than the vars field it is replacing, which heavily restricts the target fields. Has there been a conversation about what selection capabilities source/target should support? E.g. on the flip side should we be requiring sources to have kind AND name?

Reply:
Yeah, the main reason for the change was so I could reuse the getMatches function for both the target and source selectors. If we choose to add label and annotation selection support we can easily change it back. 

There hasn't yet been a in-depth conversation about what selection capabilities it should support beyond being able to have one source and multiple targets. It would be good to have a discussion about that, maybe that conversation should happen on the tracking issue https://github.com/kubernetes-sigs/kustomize/issues/3492. 

>  E.g. on the flip side should we be requiring sources to have kind AND name?
I don't see a reason that sources should be required to have kinds and names as long as the filter fails if there are multiple source matches, wdyt?

Reply:
Agree with @KnVerey  that this is more powerful than `vars`, which is fine/good.

I'd prefer to not use the `Target` struct any  more because it has two means to specify `ApiVersion`, meaning they can disagree.  This was added for the benefit of `vars`, a feature that we want to drop, so we'll likely drop or change Target too.

At the risk of splitting up this discussion too much, I added a suggested for a `Replacement` struct  to the issue discussion #3492

I think this PR need not respect the existing `ReplacementTransformer`.  We can delete that first, rather than try to keep it compiling.

Reply:
@natasha41575 agree that a Replacement can specify something that's effectively a wildcard for the source, but it has to resolve to one object or you get a build-time error.

Reply:
sounds good to me, I will update this PR accordingly. thanks!